### PR TITLE
CDAP-15030. Salesforce Streaming Source via Salesforce Streaming API

### PR DESCRIPTION
**Pull request by aonischuk**
This source tracks updates in Salesforce sObjects.
Examples of sObjects are opportunities, contacts, accounts, leads, any custom object, etc.

Salesforce Streaming API uses publish/subscribe model. Bayeux protocol and CometD are used, so that the client-to-server connection is maintained through long polling.

Client can subscribe to certain events by subscribing to what is called a pushTopic.
A pushTopic is a special sObject in salesforce DB. Example of pushTopic created via apex Code:

PushTopic pushTopic = new PushTopic();
pushTopic.Name = 'InvoiceStatementUpdates';
pushTopic.Query = 'SELECT Id, Name, Status__c, Description__c FROM Invoice_Statement__c';
pushTopic.ApiVersion = 45.0;
pushTopic.NotifyForOperationCreate = true;
pushTopic.NotifyForOperationUpdate = true;
pushTopic.NotifyForOperationUndelete = true;
pushTopic.NotifyForOperationDelete = true;
pushTopic.NotifyForFields = 'Referenced';
insert pushTopic;

The plugin can either create a PushTopic for user, or let him use an existing one.

Design doc for more details:
https://wiki.cask.co/display/CE/Salesforce+Realtime+Source

### CODE REVIEW

File: docs/SalesforceStreaming-streamingsource.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+-------------
+
+**Reference Name:** Name used to uniquely identify this source for lineage, annotating metadata, etc.
+
+**Username:** Salesforce username.
+
+**Password:** Saleforce password.
+
+**Client Id:** Application Client Id. This is also called "Consumer Key" in Salesforce UI.
+To create Client Id user needs to create a connected application in Salesforce first.
```
Comment (jayfjin) [inline docs/SalesforceStreaming-streamingsource.md]: To create a client id, the user

File: docs/SalesforceStreaming-streamingsource.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+To track the updates Salesforce PushTopic events are used.
+
+The generation of PushTopic notifications in Salesforce follows this sequence:
+1. Create a PushTopic based on a SOQL query. The PushTopic defines the channel.
+2. Clients subscribe to the channel.
+3. A record is created, updated, deleted, or undeleted (an event occurs). The changes to that record are evaluated.
+4. If the record changes match the criteria of the PushTopic query, a notification is generated by the server and received by the subscribed clients.
+
+The plugin can either create a PushTopic for you, or use an existing one.
```
Comment (jayfjin) [inline docs/SalesforceStreaming-streamingsource.md]: From now on, let's try to avoid "you" in technical documentation. I think this sentence works better as "The plugin can either create a PushTopic for the user, or use an existing one."

File: docs/SalesforceStreaming-streamingsource.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+**Username:** Salesforce username.
+
+**Password:** Saleforce password.
+
+**Client Id:** Application Client Id. This is also called "Consumer Key" in Salesforce UI.
+To create Client Id user needs to create a connected application in Salesforce first.
+
+**Client Secret:** Application Client Secret. This is also called "Consumer Secret" in Salesforce UI.
+To create Client Secret user needs to create a connected application in Salesforce first.
```
Comment (jayfjin) [inline docs/SalesforceStreaming-streamingsource.md]: same here, To create a client secret, the user

File: docs/SalesforceStreaming-streamingsource.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+**Login Url:** Salesforce OAuth2 login url.
+
+**Topic Name:** Salesforce push topic name. Plugin will track updates from this topic. If topic does
+not exist plugin creates it. To manually create pushTopic use Salesforce workbench or
```
Comment (jayfjin) [inline docs/SalesforceStreaming-streamingsource.md]: If the topic does not exist, the plugin will create it.

Comment (jayfjin) [inline docs/SalesforceStreaming-streamingsource.md]: workbench, Apex code, or API.

File: docs/SalesforceStreaming-streamingsource.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+Apex code or API.
+
+
+**Query:** Salesforce push topic query. The query is used by Salesforce to send updates to push topic.
+This field not required, if you are using an existing push topic.
```
Comment (jayfjin) [inline docs/SalesforceStreaming-streamingsource.md]: no comma needed here

File: docs/SalesforceStreaming-streamingsource.md [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+
+**Notify On Update:** Push topic property, which specifies if a update operation should generate a record.
+
+
+**Notify On Delete:** Push topic property, which specifies if an delete operation should generate a record.
+
+
+**Notify For Fields:** Push topic property, which specifies how the record is evaluated against the
+PushTopic query The NotifyForFields values are:<br>
```
Comment (jayfjin) [inline docs/SalesforceStreaming-streamingsource.md]: period after query

File: docs/SalesforceStreaming-streamingsource.md [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
++-------------+----------------------------------------------------------------------------+--------------+
+| Schema type |                              Salesforce type                               |    Notes     |
++-------------+----------------------------------------------------------------------------+--------------+
+| bool        | _bool                                                                      |              |
+| int         | _int                                                                       |              |
+| long        | _long                                                                      |              |
+| double      | _double, currency, percent                                                 |              |
+| date        | date                                                                       |              |
+| timestamp   | datetime                                                                   | Milliseconds |
```
Comment (jayfjin) [inline docs/SalesforceStreaming-streamingsource.md]: what is the difference between timestamp and time if they are both in milliseconds?

File: pom.xml [hunk 8]
```diff
@@ -280,6 +308,64 @@
...
+          <exclusion>
+            <groupId>log4j</groupId>
+            <artifactId>log4j</artifactId>
+          </exclusion>
+          <exclusion>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-client</artifactId>
+          </exclusion>
+          <exclusion>
+            <groupId>com.esotericsoftware.reflectasm</groupId>
```
Comment (jayfjin) [inline pom.xml]: what is this? do we need it?

File: src/main/java/co/cask/hydrator/salesforce/plugin/source/batch/SalesforceBatchSource.java [hunk 9]
```diff
@@ -135,19 +134,20 @@ public void transform(KeyValue<NullWritable, CSVRecord> input,
...
 
-        Schema.Field field = schema.getField(fieldName);
+        Schema.Field field = schema.getField(fieldName, true);
 
         if (field == null) {
           continue; // this field is not in schema
         }
 
-        builder.set(fieldName, convertValue(value, field));
+        builder.set(field.getName(), convertValue(value, field));
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/plugin/source/batch/SalesforceBatchSource.java]: why the change here?

File: src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+public class SalesforcePushTopicListener {
+  private static final Logger LOG = LoggerFactory.getLogger(SalesforcePushTopicListener.class);
+
+  private static final String DEFAULT_PUSH_ENDPOINT = "/cometd/" + SalesforceConstants.API_VERSION;
+  /**
+   * Timeout of 110 seconds is enforced by Salesforce Streaming API and is not configurable.
+   * So we enforce the same on client.
+   */
+  private static final int CONNECTION_TIMEOUT = 110;
+  private static final long HANDSHAKE_TIMEOUT_MS = TimeUnit.SECONDS.toMillis(110);
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java]: Comment says 110 seconds but code says 110 milliseconds. Which one is correct?

File: src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+ */
+public class SalesforcePushTopicListener {
+  private static final Logger LOG = LoggerFactory.getLogger(SalesforcePushTopicListener.class);
+
+  private static final String DEFAULT_PUSH_ENDPOINT = "/cometd/" + SalesforceConstants.API_VERSION;
+  /**
+   * Timeout of 110 seconds is enforced by Salesforce Streaming API and is not configurable.
+   * So we enforce the same on client.
+   */
+  private static final int CONNECTION_TIMEOUT = 110;
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java]: what unit is this in? Let's add to the variable name

File: src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+    } catch (Exception e) {
+      throw new RuntimeException("Could not start client", e);
+    }
+  }
+
+  /**
+   * Retrieves message from the messages queue, waiting up to the
+   * specified wait time if necessary for an element to become available.
+   *
+   * @param timeout how long to wait before giving up
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java]: what unit is timeout in? Let's include in variable name

File: src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+      }
+    };
+
+    // Now set up the Bayeux client itself
+    BayeuxClient client = new BayeuxClient(instanceUrl + DEFAULT_PUSH_ENDPOINT, transport);
+    client.handshake();
+
+    return client;
+  }
+
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java]: nit: extra newline

File: src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSource.java [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Objects;
+import javax.ws.rs.Path;
+
+/**
+ * Returns records in realtime created by Salesforce. To achieve this Salesforce Streaming API is used.
+ * We use cometd server to subscribe to Salesforce Push Topics and receive realtime updates from there
+ *
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSource.java]: nit: extra line

File: src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSource.java [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+      // In case user does not set any query in the corresponding field we need to get it from pushTopic on server
+      query = (String) pushTopic.getField("Query");
+    }
+
+    return SalesforceSchemaUtil.getSchema(authenticatorCredentials,
+                                   SObjectDescriptor.fromQuery(query));
+  }
+}
+
+
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSource.java]: nit: extra newlines

File: src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSourceConfig.java [hunk 16]
```diff
@@ -0,0 +1,297 @@
...
+  private static final long serialVersionUID = 4218063781902315444L;
+
+  protected static final String ENABLED_KEYWORD = "Enabled";
+  protected static final String PROPERTY_PUSHTOPIC_NAME = "pushTopicName";
+  protected static final String PROPERTY_PUSHTOPIC_QUERY = "pushTopicQuery";
+  protected static final String PROPERTY_SOBJECT_NAME = "sObjectName";
+
+  private static Pattern isValidFieldNamePattern = Pattern.compile("[a-zA-Z0-9.-_]+");
+
+  @Description("Salesforce push topic name. Plugin will track updates from this topic. If topic does not exist," +
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSourceConfig.java]: need a space after ","

File: src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSourceConfig.java [hunk 17]
```diff
@@ -0,0 +1,297 @@
...
+    "it will be automatically created. " +
+    "To manually create pushTopic use Salesforce workbench or Apex code or API.")
+  @Name(PROPERTY_PUSHTOPIC_NAME)
+  @Macro
+  private String pushTopicName;
+
+  @Description("Salesforce push topic query. The query is used by Salesforce to send updates to push topic." +
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSourceConfig.java]: need a space after "."

File: src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSourceConfig.java [hunk 18]
```diff
@@ -0,0 +1,297 @@
...
+    "This field not required, if you are using an existing push topic.")
+  @Nullable
+  @Name(PROPERTY_PUSHTOPIC_QUERY)
+  @Macro
+  private String pushTopicQuery;
+
+
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSourceConfig.java]: nit: extra newline

File: src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSourceConfig.java [hunk 19]
```diff
@@ -0,0 +1,297 @@
...
+  @Nullable
+  @Name("pushTopicNotifyUpdate")
+  private String pushTopicNotifyUpdate;
+
+  @Description("Push topic property, which specifies if an delete operation should generate a record.")
+  @Nullable
+  @Name("pushTopicNotifyDelete")
+  private String pushTopicNotifyDelete;
+
+
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSourceConfig.java]: nit: extra newline

File: src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSourceConfig.java [hunk 20]
```diff
@@ -0,0 +1,297 @@
...
+      SObject pushTopic = fetchPushTopicByName(partnerConnection, pushTopicName);
+      String query = getQuery();
+
+      if (pushTopic == null) {
+        LOG.info("Creating PushTopic {}", pushTopicName);
+
+        if (Strings.isNullOrEmpty(query)) {
+          throw new InvalidConfigPropertyException("SOQL query or SObject name must be provided, unless " +
+                                                     "using against existing pushTopic",
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSourceConfig.java]: "existing pushTopic is used"

File: src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSourceConfig.java [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+
+        // Ensure that pushTopic has the same parameters as user set in config.
+        // Otherwise it would be confusing for user if we continue
+        assertFieldValue(pushTopic, "NotifyForOperationCreate", isPushTopicNotifyCreate().toString());
+        assertFieldValue(pushTopic, "NotifyForOperationUpdate", isPushTopicNotifyUpdate().toString());
+        assertFieldValue(pushTopic, "NotifyForOperationDelete", isPushTopicNotifyDelete().toString());
+        assertFieldValue(pushTopic, "NotifyForFields", getPushTopicNotifyForFields());
+      }
+    } catch (ConnectionException e) {
+      throw new InvalidStageException("Cannot connect to Salesforce API with credentials specified", e);
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforceStreamingSourceConfig.java]: s/specified"/specified. "

File: src/main/java/co/cask/hydrator/salesforce/soap/SObjectUtil.java [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+import com.sforce.soap.partner.PartnerConnection;
+import com.sforce.soap.partner.SaveResult;
+import com.sforce.soap.partner.sobject.SObject;
+import com.sforce.ws.ConnectionException;
+
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Class which provides functions to work with sobjects
```
Comment (jayfjin) [inline src/main/java/co/cask/hydrator/salesforce/soap/SObjectUtil.java]: sObjects

File: src/main/java/co/cask/hydrator/salesforce/plugin/source/batch/SalesforceBatchSource.java (hunk 9 above)
Comment (aonischuk) [reply src/main/java/co/cask/hydrator/salesforce/plugin/source/batch/SalesforceBatchSource.java]: fields in salesforce are case-insensitive. And we saved them all lowercase but in this place we need to know the correct case field in CDAP schema has to be able to set the value via builder.

File: src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java (hunk 10 above)
Comment (aonischuk) [reply src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java]: It's 110 seconds. This line TimeUnit.SECONDS.toMillis(110)
just transforms 110 seconds in 110000 since we need to use milliseconds in code below. Does it make sense?

Comment (jayfjin) [reply src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java]: ah yes my bad, read the code too quickly

File: docs/SalesforceStreaming-streamingsource.md (hunk 7 above)
Comment (aonischuk) [reply docs/SalesforceStreaming-streamingsource.md]: Under the hood we are just using CDAP logical types
https://github.com/cdapio/cdap/blob/1d2d51a2e5c467891b9fe994133ef80646578a09/cdap-api-common/src/main/java/io/cdap/cdap/api/data/schema/Schema.java#L105

It's TIME_MILLIS vs TIMESTAMP_MILLIS. The one has date and time the other only date.

File: pom.xml (hunk 8 above)
Comment (aonischuk) [reply pom.xml]: The exclusions list is the same in most plugins which run streaming integration tests.
https://github.com/search?q=org%3Adata-integrations+com.esotericsoftware.reflectasm&type=Code


If we don't do exclusions integration tests runs will fail due to conflicting libs. Probably com.esotericsoftware.reflectasm depends on some libs which conflict with CDAP classes.

File: src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java (hunk 12 above)
Comment (aonischuk) [reply src/main/java/co/cask/hydrator/salesforce/plugin/source/streaming/SalesforcePushTopicListener.java]: The unit is next argument

   * @param timeout how long to wait before giving up
   * @param unit timeunit of timeout

So it can be seconds milliseconds or what ever you set it to.

**Comment by aonischuk** [comment]
@jayfjin addressed all you comments.
Can we please merge?
